modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportStartReindexJobAction.java

```diff
@@ -0,0 +1,151 @@
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.UUIDs;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.persistent.PersistentTasksCustomMetaData;
+import org.elasticsearch.persistent.PersistentTasksService;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+public class TransportStartReindexJobAction
+    extends TransportMasterNodeAction<StartReindexJobAction.Request, StartReindexJobAction.Response> {
+
+    private final PersistentTasksService persistentTasksService;
+
+    @Inject
+    public TransportStartReindexJobAction(TransportService transportService, ThreadPool threadPool,
+                                          ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver,
+                                          ClusterService clusterService, PersistentTasksService persistentTasksService) {
+        super(StartReindexJobAction.NAME, transportService, clusterService, threadPool, actionFilters, StartReindexJobAction.Request::new,
+            indexNameExpressionResolver);
+        this.persistentTasksService = persistentTasksService;
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.SAME;
+    }
+
+    @Override
+    protected StartReindexJobAction.Response newResponse() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    protected StartReindexJobAction.Response read(StreamInput in) throws IOException {
+        return new StartReindexJobAction.Response(in);
+    }
+
+    @Override
+    protected void masterOperation(StartReindexJobAction.Request request, ClusterState clusterState,
+                                   ActionListener<StartReindexJobAction.Response> listener) {
+        String generatedId = UUIDs.randomBase64UUID();
+
+        Map<String, String> filteredHeaders = threadPool.getThreadContext().getHeaders().entrySet().stream()
+            .filter(e -> ReindexPlugin.HEADER_FILTERS.contains(e.getKey()))
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+        ReindexJob job = new ReindexJob(request.getReindexRequest(), filteredHeaders);
+
+        // TODO: Task name
+        persistentTasksService.sendStartRequest(generatedId, ReindexTask.NAME, job, new ActionListener<>() {
+                @Override
+                public void onResponse(PersistentTasksCustomMetaData.PersistentTask<ReindexJob> persistentTask) {
+                    if (request.getWaitForCompletion()) {
+                        waitForReindexDone(persistentTask.getId(), listener);
```

Comment:
As we discussed, this part can be left out of this for now, since if coordinator fails, the request would also fail. The important part is that reindex continues (which it does).